sql/binlog.cc

```diff
@@ -5013,14 +5013,20 @@ bool MYSQL_BIN_LOG::check_write_error(THD *thd) {
 
   if (!thd->is_error()) DBUG_RETURN(checked);
 
-  switch (thd->get_stmt_da()->mysql_errno()) {
-    case ER_TRANS_CACHE_FULL:
-    case ER_STMT_CACHE_FULL:
-    case ER_ERROR_ON_WRITE:
-    case ER_BINLOG_LOGGING_IMPOSSIBLE:
-      checked = true;
-      break;
+  // Check all conditions for one that matches the expected error
```

Comment:
This is a known upstream bug
https://bugs.mysql.com/bug.php?id=88223
that we fixed in Percona Server
https://jira.percona.com/browse/PS-1107
https://github.com/percona/percona-server/pull/2574 (Pull Request for 8.0)
Please, consider taking MTR test cases from there.